docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,343 @@
+
+We have to make a choice between:
+
+- new behavior by default, `UPGRADE.md` tells users to set a variable to keep
+  old behavior
+- old behavior by default, new users have to be convinced to set a variable to
+  get new behavior
+
+### `MeshService`-targeted policy
+
+If for example we have a policy targeted at:
+
+```
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: demo-app_kuma-demo_svc_5000
+```
+
+when switching over to MeshService, this policy will no longer apply.
+
+We can always convert the above to an additional entry:
+
+```
+  - targetRef:
+      kind: MeshService
+      labels:
+        kuma.io/display-name: demo-app
+        kuma.io/namespace: kuma-demo
+      port: 5000
+```
+
+Potentially we could offer a tool to convert these refs from old-style to
+new-style `MeshService`.
+
+#### Require `sectionName`
+
+Because the original refers to instances of `demo-app_kuma-demo_svc_5000` from
+any zone, we have to use `labels` to match instances outside of the local
+zone.
+
+Note that there's an ambiguity here with universal. We can't tell whether:
+
+```
+spec:
+  to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+```
+
+is meant to target `kuma.io/service` or a `MeshService` named `backend` and thus
+we can't determine whether all instances of `kuma.io/service: backend` are meant
+or only the instances matched by `MeshService`.
+
+Thus we have to require `sectionName` to a port, in order to refer to
+`MeshServices` unambiguously.
+
+#### Port name
+
+We can't require users to add a `name` field to their existing `Services` or
+annotate their `Dataplanes` for universal, so we should
+set a canonical name for every port assuming one isn't set.
+
+Because the `port` field must be unique anyway, this could be a suitable,
+automatic `name`.
+
+### Stats
+
+For the Kuma API it's important that we can derive the destination that corresponds
+to a given metric label, like `name_namespace_svc_port`.
+In particular, it's valuable if we can do this without making additional
+requests.
+
+If we have a label like `name_namespace_svc_port` but we need to make a request
+in order to figure out what resource it correponds to, that's a disadvantage.
+
+### Traffic
+
+At the moment, we generate a VIP for every `kuma.io/service` and a hostname
+`<kuma.io/service>.mesh` that points to this VIP. The VIP leads to an outbound
+listener and a cluster being created.
+
+#### Before `MeshService` we have
+
+- One cross-zone cluster for a `kuma.io/service`
+
+- A `<kuma.io/service>.mesh` listener for Kuma generated VIP -> cross-zone cluster
+
+On Kubernetes we also have
+
+- `name.namespace.svc.cluster.local` -> ClusterIP -> same cross-zone cluster
+
+#### Once `MeshService` is created
+
+`MeshService` takes over _the cluster generation_ but does
+not affect the VIP generation. So we now have:
+
+- `<kuma.io/service>.mesh` -> Kuma VIP -> local-zone cluster
+- `name_namespace_svc_port` -> ClusterIP -> local-zone cluster
```

Comment:
I don't really understand this, what do you mean by `name_namespace_svc_port` here? it's not hostname right?

Reply:
yeah this is meant to be `name.namespace.cluster.local`